File: 2021/src/solutions/Day1.java
package solutions;

import java.util.List;

import static utils.Utils.*;

/*
Result:

      -------Part 1--------   -------Part 2--------
Day       Time  Rank  Score       Time  Rank  Score
  1   00:01:40   407      0   00:09:07  2101      0
 */
public class Day1 {

    public static void main(String[] args) throws Exception {
        List<String> input = day(1);
        List<String> sample1 = sample("""
                199
                200
                208
                210
                200
                207
                240
                269
                260
                263
                """);

        System.out.println("Sample1: " + part1(sample1));
        System.out.println("Result: " + part1(input));

        System.out.println("---");

        System.out.println("Sample1: " + part2(sample1));
        System.out.println("Result: " + part2(input));
    }

    static long part1(List<String> strings) throws Exception {
        long[] longs = longs(strings);

        long result = 0;
        long prev = -1;
        for (int i = 0; i < longs.length; i++) {
            long d = longs[i];
            if (prev != -1) {
                result += d > prev ? 1 : 0;
            }
            prev = d;
        }

        return result;
    }

    static long part2(List<String> strings) throws Exception {
        long[] longs = longs(strings);

        long result = 0;
        long[] windows = new long[longs.length - 2];
        for (int i = 0; i < longs.length - 2; i++) {
            windows[i] = longs[i] + longs[i + 1] + longs[i + 2];
        }
        long prev = -1;
        for (int i = 0; i < windows.length; i++) {
            if (prev != -1) {
                result += windows[i] > prev ? 1 : 0;
            }
            prev = windows[i];
        }

        return result;
    }

}

File: 2021/src/solutions/Day1.js
#!/usr/bin/env node

let utils = require("../utils/utils");

function main() {
    let input = utils.day(1);
    let sample1 = utils.sample(`
        199
        200
        208
        210
        200
        207
        240
        269
        260
        263
    `)

    console.log(`Sample1: ${part1(sample1)}`)
    console.log(`Result: ${part1(input)}`)

    console.log("---")

    console.log(`Sample1: ${part2(sample1)}`)
    console.log(`Result: ${part2(input)}`)
}

function part1(strings) {
    let longs = utils.longs(strings);

    let result = 0;
    let prev = longs[0];
    for (let i = 1; i < longs.length; i++) {
        if (longs[i] > prev) {
            result++;
        }
        prev = longs[i];
    }

    return result;
}

function part2(strings) {
    let longs = utils.longs(strings);

    let windows = [];
    for (let i = 0; i < longs.length; i++) {
        windows.push(longs[i] + longs[i+1] + longs[i+2]);
    }
    let result = 0;
    let prev = windows[0];
    for (let i = 1; i < windows.length; i++) {
        if (windows[i] > prev) {
            result++;
        }
        prev = windows[i];
    }

    return result;
}

main();

